SQL: Introduce an async querying mode for SQL

bpintea opened the pull request:
This adds an async query mode to SQL.
It (re)uses the same request and response async-specific EQL object
parameters.

Also similar to EQL, the running search task can have its state
monitored and canceled and its results stored and deleted, with
intermediary responses not supported (the entire result is available
once search finished).

The initial query and subsequent pagination/scrolling requests will both
be started in the async mode.

bpintea committed:
Introduce an async querying mode for SQL

This adds an async query mode to SQL.
It (re)uses the same request and response async-specific EQL object
parameters.

Also similar to EQL, the running search task can have its state
monitored and canceled and its results stored and deleted, with
intermediary responses not supported (the entire result is available
once search finished).

The initial query and subsequent pagination/scrolling requests will both
be started in the async mode.

bpintea commented:
Note: text formatting not yet supported, that will be applied with a [subsequent PR](https://github.com/elastic/elasticsearch/pull/72166/commits/d51eb98d347d20aaf1d8b5a542db09a9070e2b75).

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

imotov commented on x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java:
```diff
@@ -300,6 +303,7 @@
         "cluster:monitor/xpack/searchable_snapshots/stats",
         "cluster:monitor/xpack/security/saml/metadata",
         "cluster:monitor/xpack/spatial/stats",
+        SQL_ASYNC_GET_STATUS_ACTION_NAME,
```
I am really curious why we don't use constants for all other actions here. @ywangd or @tvernum could you clarify?

ywangd replied:
The main reason is that the list is generated when the test is first introduced, i.e. the test reports all missing names when this list is empty. Also a few other reasons:
* Many names are not available to this test, e.g. autoscaling, enrich, grok processor, reindex
* Some names are computed, e.g. all the xpack usage actions and shard level actions (`[s]`) and quite a few others
* I find a sorted list of consistent literal strings easier to read and useful to scan through
* Sometimes it is useful to repeat the literal string in tests which can itself act as a test.

imotov replied:
So, it sounds like we have several good enough reasons to continue using strings here instead of constants, and we should probably continue doing it in this case.

bpintea replied:
Replaced the var with the string (and added the var location as a comment).

imotov commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/async/StoredAsyncResponse.java:
```diff
@@ -5,7 +5,7 @@
  */
 
-package org.elasticsearch.xpack.eql.async;
+package org.elasticsearch.xpack.ql.async;
```
I would keep this generic and move it into org.elasticsearch.xpack.core.async so ML could use it. The same for StoredAsyncTask and

bpintea replied:
Thanks, @imotov!
Will wait one more review and depending on the amount of changes required will move the classes either part of this, or of a next PR that I have to open anyways.

bpintea replied:
The two classes have been moved.

imotov commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/async/StoredAsyncTask.java:
```diff
@@ -95,4 +95,15 @@ protected synchronized void onFailure(Exception e) {
         taskManager.cancelTaskAndDescendants(this, reason, true, ActionListener.wrap(runnable));
     }
+
+    public static QlStatusResponse getStatusResponse(StoredAsyncTask<?> asyncTask) {
```
I would introduce it one level above so we can keep this task generic and usable by ML.

imotov reviewed (approved):
LGTM in general. I think we can move some classes that we moved to `xpack.ql.async` all the way to `xpack.core.async` since ML expressed desire to also use them.

matriv commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/async/QlStatusResponse.java:
```diff
@@ -4,40 +4,48 @@
-            Long startTimeMillis,
-            long expirationTimeMillis,
-            RestStatus completionStatus) {
+    public interface AsyncStatus {
```
Imho, I would extract this to its own file.

bpintea replied:
I find the interface tightly bound to the outer class (i.e. no independent applicability), that's why I was also using it as a prefix wherever the interface is used. But if you think it'd be better practice to extract it, I can still rename it and do it.

(Thanks for this long-PR review!)

matriv replied:
I'll leave it up to you, not insisting, it's just that from the name `AsyncStatus` it seems like a generic iface and could also belong to a separate file.

matriv commented on x-pack/plugin/sql/qa/server/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/RestSqlSecurityAsyncIT.java:
```diff
@@ -0,0 +1,176 @@
+        testCase("user2", "user1");
+    }
+
+    private void testCase(String user, String other) throws Exception {
```
nit, to make it more clear:
```suggestion
    private void testCase(String user, String otherUser) throws Exception {
```

bpintea replied:
Generally I've tried to keep as close to the original EQL tests as possible (since many of them can't be deduplicated EQL-SQL). But this is small, I'll change it.

matriv reviewed (commented):
LGTM, Nice work! Left a couple of minor comments.

bpintea committed:
Merge remote-tracking branch 'upstream/feat/sql-async' into feat/ql_async

bpintea committed:
Address review comments

- move StoredAsyncResponse and StoredAsyncTask classes from ql to
core.async packages.
- rename function parameter.

bpintea committed:
Adress review comment

Replace constant with hardcoded string.

mark-vieira commented:
jenkins re-test this please

astefan reviewed (approved):
LGTM with some minor comments.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -144,6 +145,10 @@ public void query(List<Attribute> output, QueryContainer query, String index, Ac
             l = new ScrollActionListener(listener, client, cfg, output, query);
         }
 
+        if (cfg.task() != null && cfg.task().isCancelled()) {
```
There are a couple of places where the task is checked against being canceled within SQL, both placed before "leaving" SQL:
* before resolving the index;
* here, before running the search.

It makes sense to do it there because (1) compared to the distributed part, the SQL code will run quick; and (2) there's anyways a race between the cancelation and task execution -- one could add more checkpoints throughout SQL, but it won't practically improve much (like reactivity to the cancelation).

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/async/QlStatusResponse.java:
```diff
@@ -109,15 +117,17 @@ public RestStatus status() {
+                builder.timeField("start_time_in_millis", "start_time", startTimeMillis);
+            }
+            builder.timeField("expiration_time_in_millis", "expiration_time", expirationTimeMillis);
+            if (isRunning == false) { // completion status is available only for a completed eql search
```
Thanks, will fix them with the following PR.
